semanticdb/metacp/src/main/scala/scala/meta/cli/Metacp.scala

```diff
@@ -1,25 +1,94 @@
 package scala.meta.cli
 
 import java.io._
+import java.nio.file.Files
+
+import io.github.soc.directories.ProjectDirectories
+import org.langmeta.internal.io.PlatformFileIO
+import org.langmeta.io.AbsolutePath
+import org.langmeta.io.Classpath
+
 import scala.meta.internal.metacp._
 
 object Metacp {
   def main(args: Array[String]): Unit = {
     sys.exit(process(args, System.out, System.err))
   }
 
-  @deprecated("Use `process(args, System.out, System.err)` instead.", "3.4.0")
   def process(args: Array[String]): Int = {
     process(args, System.out, System.err)
   }
 
   def process(args: Array[String], out: PrintStream, err: PrintStream): Int = {
-    Settings.parse(args.toList) match {
+    Settings.parse(args.toList, out, err) match {
       case Some(settings) =>
-        val main = new Main(settings, out, err)
-        main.process()
+        val mclasspath = process(settings)
+        out.println(mclasspath.mkString(File.pathSeparator))
+        0
       case None =>
         1
     }
   }
+
+  def process(settings: Settings): List[AbsolutePath] = {
+    settings.classpath.shallow.toParArray.map(p => processPath(p, settings)).toList
+  }
+
+  def processPath(in: AbsolutePath, settings: Settings): AbsolutePath = {
```

Comment:
I think we have way too many public APIs here. Instead of `main/process/process/process/processPath`, I suggest that we have just two methods:
  * `def main(args: Array[String]): Unit` (this one we can't avoid)
  * `def process(settings: Settings, reporter: Reporter): Result` (where `Result` carries both an indication of success and the list of absolute paths).

Sure, this will be less convenient to clients, because they will have to manually construct and deconstruct the API structures to use `process`, but this way we keep the public API minimal.

Reply:
Implemented in https://github.com/scalameta/scalameta/pull/1394.